### ISSUE TITLE
Support for Voicemeeter

### ISSUE THREAD

**Issue by root-reindeer-flotilla** [jaakkopasanen/AutoEq#682]
Would support for Voicemeeter/Voicemeeter Banana EQ be possible? It's only a 6 segment and the lowest on autoeq is 10.

https://vb-audio.com/Voicemeeter/banana.htm

Thanks! :)

**Comment by ecker00** [comment]
Looking for the same thing, but I don't see any way to import/export the EQ from the UI. It has a copy/paste button, but doesn't seem to copy anything outside the Voicemeeter interface. So even if adding support, seems that Voicemeeter might have to add an import/export feature, or let AutoEq show the settings so we can plot it in manually.

**Comment by JuhaTP** [comment]
I've coded a routine which writes EQ settings for Voicemeeter Banana EQ. I'm not familiar with GitHub features so I give all needed here if someone likes to pull request.

New function to be added in frequency_response.py
 
def write_voicemeeter_peq(self, file_path, peqs):
        file_path = os.path.abspath(file_path)
        f = self.generate_frequencies(f_step=DEFAULT_BIQUAD_OPTIMIZATION_F_STEP)
        compound = PEQ(f, peqs[0].fs, [])
        for peq in peqs:
            for filt in peq.filters:
                compound.add_filter(filt)

        EType = 0
        with open(file_path.replace('.txt', '.xml'), 'w', encoding='utf-8') as f:
            s = f'<?xml version="1.0" encoding="utf-8"?>\n'
            s += f'<VBAudioVoicemeeterBUSEQConfig>\n'
            s += f'<VoiceMeeterBUSEQ>\n'
            for idx in range(1,2):
                for channel in range(1,9):
                    for i, filt in enumerate(compound.filters):
                        s += f'\t<Bus index=\'{idx:d}\' ' 
                        s += f'channel=\'{channel:d}\' '
                        s += f'cell=\'{i+1:d}\' '
                        s += f'EQon=\'1\' ' 
                        if i > 0 and i < 5:
                            EType = 0
                        elif i == 0:
                            EType = 5
                        else:
                            EType = 6
                        s += f'EQtype=\'{EType:d}\' ' 
                        s += f'dblevel=\'{filt.gain:.2f}\' '
                        s += f'freq=\'{filt.fc:.2f}\' '
                        s += f'Q=\'{filt.q:.2f}\' />\n'
                            #print(s)
                            
            s += f'</VoiceMeeterBUSEQ>\n'
            s += f'</VBAudioVoicemeeterBUSEQConfig>'
            f.write(s)

Changes/additions in EutoEq source codes:

**__main__.py**

```
43 arg_parser.add_argument('--voicemeeterpeq', action='store_true',
                            help='Will produce SuperEQ settings if this parameter exists, no value needed.')  
``` 
                            
**batch_processing.py**

```
19 def batch_processing(..., voicemeeterpeq=False, ...)

33 if not compensation and (... or voicemeeterpeq, ...) 
  
110 args = (..., voicemeeterpeq, ...)

133 def process_file(..., voicemeeterpeq, ...)

~160     if voicemeeterpeq:
        	voicemeeterpeq = True
~235 if voicemeeterpeq:
        parametric_peqs = fr.optimize_parametric_eq(
            parametric_eq_config, fs[0], preamp=preamp) if parametric_eq else None
        fr.write_voicemeeter_peq(output_file_path.replace('.csv', '.xml'), parametric_peqs)
     else:
        parametric_peqs = None
```

**constants.py**

     ```
'VOICEMEETER_PEQ': {
        'optimizer': {
            'min_f': 20,
            'max_f': 20000,
            'max_time': None,
            'min_change_rate': None,
            'min_std': 0.008,
        },
        'filter_defaults': {
            'min_fc': 20.0,
            'max_fc': 10000.0,
            'min_q': 0.18248,
            'max_q': 6.0,
            'min_gain': -20,
            'max_gain': 20
        },
        'filters': [{
            'type': 'LOW_SHELF',
            'fc': 65.0,
            'q': 0.7
        }, 
            {'type': 'PEAKING'},
            {'type': 'PEAKING'},
            {'type': 'PEAKING'},
            {'type': 'PEAKING'},
        {
            'type': 'HIGH_SHELF',
            'fc': 16744.036,
            'min_fc': 15000,
            'max_fc': 20000,
            'min_q': 0.4,
            'max_q': 0.7
        }] 
    }
```
    
**frequency_response.py**
    
```
38 	def __init__(..., voicemeeterpeq=None, ...)
	...
	self.voicemeeterpeq = self._init_data(voicemeeterpeq)
	...

~75  def copy(..., voicemeeterpeq=self._init_data(self.voicemeeterpeq),...)

~135   def reset(..., voicemeeterpeq=True, ...)
	...
	        if voicemeeterpeq:
            self.voicemeeterpeq = self._init_data(None)
        ...

~490  def compensate(
	...
	self.reset(...,  voicemeeterpeq=True, ...)
	
~1080 def smoothen_fractional_octave(
	...
	self.reset(..., voicemeeterpeq=True, ...)
```
I have tested functionality  with command line script:
	
`python3 -m autoeq --input-file="measurements/oratory1990/data/over-ear/AKG K240 Studio.csv" --output-dir="my_results" --bass-boost 3 --compensation="compensation/harman_over-ear_2018_wo_bass.csv" --voicemeeterpeq --parametric-eq --parametric-eq-config=VOICEMEETER_PEQ`

which produces **AKG K240 Studio.xml** :

<?xml version="1.0" encoding="utf-8"?>
<VBAudioVoicemeeterBUSEQConfig>
<VoiceMeeterBUSEQ>
	<Bus index='1' channel='1' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='1' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='1' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='1' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='1' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='1' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='2' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='2' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='2' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='2' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='2' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='2' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='3' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='3' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='3' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='3' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='3' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='3' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='4' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='4' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='4' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='4' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='4' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='4' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='5' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='5' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='5' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='5' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='5' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='5' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='6' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='6' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='6' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='6' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='6' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='6' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='7' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='7' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='7' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='7' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='7' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='7' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
	<Bus index='1' channel='8' cell='1' EQon='1' EQtype='5' dblevel='6.09' freq='65.00' Q='0.70' />
	<Bus index='1' channel='8' cell='2' EQon='1' EQtype='0' dblevel='-3.37' freq='205.49' Q='0.40' />
	<Bus index='1' channel='8' cell='3' EQon='1' EQtype='0' dblevel='5.36' freq='4220.58' Q='2.96' />
	<Bus index='1' channel='8' cell='4' EQon='1' EQtype='0' dblevel='3.52' freq='71.50' Q='0.93' />
	<Bus index='1' channel='8' cell='5' EQon='1' EQtype='0' dblevel='5.50' freq='1591.25' Q='3.11' />
	<Bus index='1' channel='8' cell='6' EQon='1' EQtype='6' dblevel='3.70' freq='16744.04' Q='0.70' />
</VoiceMeeterBUSEQ>
</VBAudioVoicemeeterBUSEQConfig>

![AKG K240 Studio](https://github.com/jaakkopasanen/AutoEq/assets/141587285/925c16fd-e194-4dc2-bf40-f907eedb1210)

Also tested with a config file voicemeeterpeq.yaml :

optimizer:
  min_f: 20
  max_f: 20000  # Only optimizing up to 10 kHz because of we only have 6 filters
  max_time: null  # Letting optimizer finish since this is probably for a one-off custom optimization, not for API
  min_change_rate: null  # Letting optimizer finish
  min_std: null  # Letting optimizer finish
filter_defaults:
  min_fc: 20
  max_fc: 10000
  min_q: 0.18248
  max_q: 6
  min_gain: -20
  max_gain: 20
filters:
  - type: LOW_SHELF
    fc: 65  # Fixed center frequency at 105 Hz
    q: 0.7  # Fixed quality of 0.7
  - type: PEAKING
  - type: PEAKING
  - type: PEAKING
  - type: PEAKING
  - type: HIGH_SHELF
    fc: 9000
    min_fc: 5000  # Letting the optimizer adjust high shelf between 5 kHz and...
    max_fc: 12000  # ...12 kHz
    min_q: 0.4  # Shelf filters start to overshoot below 0.4
    max_q: 0.7  # Shelf filters start to overshoot after 0.7

which generates similar .xml file (some differences in parameter values but response is about the same):

`python3 -m autoeq --input-file="measurements/oratory1990/data/over-ear/AKG K240 Studio.csv" --output-dir="my_results" --bass-boost 3 --compensation="compensation/harman_over-ear_2018_wo_bass.csv" --voicemeeterpeq --parametric-eq --parametric-eq-config=voicemeeterpeq.yaml`

Both .xml files loaded properly in Voicemeeter BANANA.

**Comment by jaakkopasanen** [comment]
Voicemeeter added as an eq app option in the webapp.
